### PR TITLE
Kafka producer consumer setup

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Issue | Related to https://github.com/ripe-tech/ripe-registry/issues/30 |
| Decisions | Initial setup for the repo, with API for kafka producer and consumer. The setup of the Kafka client and respective producer and consumer is made using `yonius.json`, and a consumer and producer is created. The consumer subscribes to the given topics, and the producer produces to the topic given when the produce request is made. <br><br> **Consumer**: <br> • Receives a callback function that will be called for each new message. It implements default onSuccess and onError logic, which can be overrided by functions `onSuccess` and `onError` in the `options`. <br> • The consumer is responsible for the retry logic by using a retry buffer, and will try the given number of retries until they expire. When that happens, the `sendError` method is called. br> • The consumer persists the failed messages while they still have retries. This persistence is made in the JSON file, which is updated when a new retry is made, a new message fails or a message expires their retries. <br> • To use the consumer: `bus.bind(<topic>, (message) => {...})` or `bus.bind(<topic>, { callback: (message) => {...}, ... })`<br><br> **Producer**: <br> • Sends an array of messages given, which will be converted to strings if they are provided as objects.  <br> • To use the producer: `bus.trigger(<topic>, <message>)`|
| Animated GIF | -- |

### CODE REVIEW

File: js/consumer.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
+import { conf } from "yonius";
+import * as fs from "fs";
+
+console.log(fs);
```
Comment (joamag) [inline js/consumer.js]: ?

File: js/consumer.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
+import { conf } from "yonius";
+import * as fs from "fs";
+
+console.log(fs);
+
+export class Consumer {
```
Comment (joamag) [inline js/consumer.js:3]: Let's use the adapter pattern KafkaConsumer that respects the base consumer API

File: js/producer.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { conf } from "yonius";
+import { CompressionTypes } from "kafkajs";
+
+export class Producer {
```
Comment (joamag) [inline js/producer.js:3]: Same here, `KafkaProducer` that respects a common `Producer` API

**Review by joamag — CHANGES_REQUESTED** [review]

**Comment by joamag** [comment]
@BeeMargarida on top of these changes an upper layer (much more abstract) for producing events should exist where the inner details (connection, etc.) are abstracted from the developer so that a producer node could just do something like

```javascript
bus.trigger("order.create", {order_number: "32132", ...})
```

all the singleton client-related logic is abstract by this API and the first time a trigger occurs a connection is established via the currently active engine and using the configuration details

**Comment by joamag** [comment]
For the upper layer consumer API the equivalent: 

```javascript
bus.bind("order.created", (order) => {
});
```

**Comment by BeeMargarida** [comment]
@joamag I implemented a default `onSuccess` and `onError` logic, which can be overwritten by the options object in the `bind` and `trigger` methods. Regarding errors in the callback of bind, I changed the implementation to a try catch.

I have a doubt however, regarding the parameters for the Kafka client, consumer and producer. For now they are using the values in the `yonius.json`, but would it be better to allow them to be overridden by a `kwargs` in the API constructor?

File: README.md [hunk 4]
```diff
@@ -1,3 +1,20 @@
...
+import { API as RipeBusAPI } from "ripe-bus-api-js";
+
+await RipeBusAPI.load();
+const bus = new RipeBusAPI();
+```
+
+## Usage
+
+```javascript
+bus.bind("order", (message) => { ... })
```
Comment (joamag) [inline README.md]: separate bind from trigger and make the second parameter either a function or an Object

File: js/kafkaClient.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { conf } from "yonius";
+import { Kafka } from "kafkajs";
+
+export const KafkaClient = (function() {
+    var client = null;
```
Comment (joamag) [inline js/kafkaClient.js]: var ?

File: js/kafkaConsumer.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    _readFromBufferFile() {
+        const dir = conf("KAFKA_CONSUMER_RETRY_PERSISTENCE_DIR", "data");
+        if (!fs.existsSync(dir)) {
+            fs.mkdirSync(dir);
+            return;
+        }
+
+        if (!fs.existsSync(`${dir}/retry.json`)) return;
+        const data = fs.readFileSync(`${dir}/retry.json`, { encoding: "utf-8" });
```
Comment (joamag) [inline js/kafkaConsumer.js]: use the async fs api

Comment (joamag) [reply js/kafkaConsumer.js]: https://nodejs.org/api/fs.html#fs_fs_promises_api

File: js/kafkaConsumer.js [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+                    if (!options.autoConfirm) continue;
+                    if (options.onSuccess) options.onSuccess(parsedMessage);
+                    else this._onSuccess(message);
+                }
+            }
+        });
+    }
+
+    async _retry(callback, options = {}) {
+        if (this.retryBuffer.length === 0) return;
```
Comment (joamag) [inline js/kafkaConsumer.js]: encapsulate this around a `KafkaRetryConsumer` abstraction that inherits from `KafkaConsumer`

File: js/kafkaProducer.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    async disconnect() {
+        await this.producer.disconnect();
+    }
+
+    async produce(topic, messages, options = {}) {
+        const convertedMessages = this._convertMessages(messages);
+
+        await this.producer.send({
+            topic: topic,
+            acks: conf("KAFKA_PRODUCER_ACKS", 0),
```
Comment (joamag) [inline js/kafkaProducer.js]: try to make this as part of the constructor logic

File: js/kafkaProducer.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
+import { conf } from "yonius";
+import { CompressionTypes } from "kafkajs";
+import { Producer } from "./producer";
+import { KafkaClient } from "./kafkaClient";
+
+export class KafkaProducer extends Producer {
+    constructor() {
+        super();
+
+        const kafkaClient = KafkaClient.getInstance();
```
Comment (joamag) [inline js/kafkaProducer.js]: `options.acks === undefined ? conf("KAFKA_PRODUCER_ACKS", trure) : options.acks`

File: js/kafkaProducer.js (hunk 8 above)
Comment (joamag) [reply js/kafkaProducer.js]: `acks: options.acks === undefined ? this.acks : options.acks`

**Review by joamag — CHANGES_REQUESTED** [review]

**Review by joamag — APPROVED** [review]

**Comment by joamag** [comment]
👍

**Pull request merged**